### PR TITLE
Add support for concurrent Miri

### PR DESCRIPTION
**Pull request by vakaras**
This PR includes two changes:

1. Update Priroda to work with the newest version of Miri.
2. Change the `step` function to work with the updated interface of concurrent Miri.

This PR depends on https://github.com/rust-lang/miri/pull/1284

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -136,7 +138,9 @@ fn create_ecx<'tcx>(tcx: TyCtxt<'tcx>) -> InterpCx<'tcx> {
             ignore_leaks: true,
             seed: None,
             tracked_pointer_tag: None,
+            tracked_alloc_id: None,
             validate: true,
+            stacked_borrows: true,
```
Comment (bjorn3) [inline src/main.rs]: I am not sure stacked borrows interacts well with the locals display.

File: src/step.rs [hunk 2]
```diff
@@ -96,45 +97,48 @@ where
...
-                            && !pcx.config.bptree.is_at_breakpoint(&pcx.ecx)
-                        {
-                            continue;
-                        }
-                    } else {
-                        // Unwinding, but no cleanup for this frame
-                        // FIXME make step behaviour configurable
+    let ret: InterpResult<'_, _> = (|| {
+        loop {
+            if pcx.ecx.stack().len() <= 1 && is_ret(&pcx.ecx) {
```
Comment (bjorn3) [inline src/step.rs]: This would now prevent any thread from exiting. Please only do this when there is only one thread left.

**Review by bjorn3 — CHANGES_REQUESTED** [review]
The rustup changes LGTM. For the thread support it would be nice to have a way to show the stack for every thread and if a thread is blocking, what it is blocking on.

**Comment by bjorn3** [comment]
@oli-obk It seems that the Travis CI integration broke.

**Comment by oli-obk** [comment]
works on master... cycling the PR

**Comment by bjorn3** [comment]
This will need to change `rust-toolchain` once the rustc PR is merged.

**Comment by bjorn3** [comment]
@vakaras https://github.com/rust-lang/miri/pull/1284 got merged. Can you update this PR?

**Comment by bjorn3** [comment]
Friendly ping @vakaras.

**Comment by vakaras** [comment]
Sorry, for taking so long to update this. I am a bit swamped with other things right now.

**Comment by vakaras** [comment]
> The rustup changes LGTM. For the thread support it would be nice to have a way to show the stack for every thread and if a thread is blocking, what it is blocking on.

I agree that this would be nice, but, unfortunately, this would be a more substantial change and right now I do not have enough time for that.

**Comment by oli-obk** [comment]
> I agree that this would be nice, but, unfortunately, this would be a more substantial change and right now I do not have enough time for that.

that's totally fine in my book. We can add additional logic for showing the stack of other threads than the current one later. Let's just get the general logic working

File: src/step.rs [hunk 3]
```diff
@@ -96,57 +100,73 @@ where
     F: Fn(&InterpCx) -> ShouldContinue,
 {
     let mut message = None;
-    loop {
-        if pcx.ecx.stack().len() <= 1 && is_ret(&pcx.ecx) {
```
Comment (bjorn3) [inline src/step.rs:100]: The new code is not equivalent to the old code. The old code would prevent stepping when doing so would exit the program. The new code does allow stepping out of the `main` function.

File: src/step.rs [hunk 4]
```diff
@@ -96,57 +100,73 @@ where
...
-                if pcx.config.bptree.is_at_breakpoint(&pcx.ecx) {
+                SchedulingAction::Stop => {
+                    message = Some("interpretation finished".to_string());
                     break;
                 }
             }
-            Ok(false) => {
-                message = Some("interpretation finished".to_string());
+            let info = pcx.ecx.preprocess_diagnostics();
+            assert!(pcx.ecx.step()?, "a terminated thread was scheduled for execution");
```
Comment (bjorn3) [inline src/step.rs]: Whenever possible please report any unexpected things to the user and continue. It shouldn't just crash on unexpected events, as that would make it much less useful for debugging miri errors.

File: src/step.rs [hunk 5]
```diff
@@ -96,57 +100,73 @@ where
...
-                            && crate::should_hide_stmt(&blck.statements[frame.stmt])
-                            && !pcx.config.bptree.is_at_breakpoint(&pcx.ecx)
-                        {
-                            continue;
-                        }
-                    } else {
-                        // Unwinding, but no cleanup for this frame
-                        // FIXME make step behaviour configurable
-                    }
+    let ret: InterpResult<'_, _> = (|| {
```
Comment (bjorn3) [inline src/step.rs]: You could use a `try` block.

**Review by bjorn3 — DISMISSED** [review]

**Comment by bjorn3** [comment]
> > I agree that this would be nice, but, unfortunately, this would be a more substantial change and right now I do not have enough time for that.
> 
> that's totally fine in my book. We can add additional logic for showing the stack of other threads than the current one later. Let's just get the general logic working

Opened #25

File: src/step.rs (hunk 3 above)
Comment (vakaras) [reply src/step.rs:100]: Oh… Thanks for explaining what this piece of code is supposed to do! ☺

File: src/step.rs (hunk 5 above)
Comment (vakaras) [reply src/step.rs]: Do you mean this: https://doc.rust-lang.org/nightly/unstable-book/language-features/try-blocks.html? Or do you have some other feature in mind?

Comment (bjorn3) [reply src/step.rs]: Yes, that feature. It is used elsewhere in priroda too.

Comment (vakaras) [reply src/step.rs]: Thanks! I never used this feature before.

**Review by oli-obk — APPROVED** [review]

File: src/step.rs [hunk 6]
```diff
@@ -96,57 +100,82 @@ where
...
+                }
+            }
+            if let ShouldContinue::Stop = continue_while(&pcx.ecx) {
+                break;
+            }
+            if pcx.config.bptree.is_at_breakpoint(&pcx.ecx) {
                 break;
             }
         }
+        ()
```
Comment (bjorn3) [inline src/step.rs]:
```suggestion
```

I don't think this is necessary.

**Comment by oli-obk** [comment]
Thanks so much for doing this and bearing with us through the changes!

**Comment by vakaras** [comment]
No problem and sorry that it took so long for me to come back to this.

**Pull request merged**